dotnet/src/InternalUtilities/src/Data/VectorStoreRecordPropertyReader.cs

```diff
@@ -337,6 +338,61 @@ public string GetJsonPropertyName(string dataModelPropertyName)
         return this._jsonPropertyNamesMap.Value[dataModelPropertyName];
     }
 
+    /// <summary>
+    /// Get the <see cref="VectorStoreRecordVectorProperty"/> for the given property name.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">When there is no match.</exception>
+    public VectorStoreRecordVectorProperty GetVectorProperty<TRecord>(VectorSearchOptions<TRecord>? searchOptions)
+    {
+        if (searchOptions is not null)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            string? propertyName = searchOptions.VectorPropertyName;
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            if (searchOptions.VectorProperty is Expression<Func<TRecord, object?>> expression)
+            {
+                if (expression.Body is ConstantExpression constant && constant.Value is string text && !string.IsNullOrWhiteSpace(text))
+                {
+                    // This is something that VectorStoreGenericDataModel can use.
+                    propertyName = text;
+                }
+                // r => r.PropertyName is translated into
+                // r => (object)r.PropertyName
+                else if (expression.Body is UnaryExpression unary
+                    && unary.Operand.NodeType == ExpressionType.MemberAccess
+                    && unary.Operand is MemberExpression member
+                    && expression.Parameters.Count == 1
+                    && member.Expression == expression.Parameters[0]
+                    && member.Member is PropertyInfo property)
+                {
+                    for (int i = 0; i < this.VectorPropertiesInfo.Count; i++)
+                    {
+                        if (this.VectorPropertiesInfo[i] == property)
+                        {
+                            return this.VectorProperties[i];
+                        }
+                    }
+
+                    throw new InvalidOperationException($"The property {property.Name} of {typeof(TRecord).FullName} is not a Vector property.");
+                }
+                else
```

Comment:
Unless I'm mistaken, this doesn't support the generic data model, since the vector properties there are in a dictionary, and so the user would do a dictionary lookup in the property selector - we'll need to support that too. 

Note that the generic data model also isn't supported in LINQ filter querying yet (#10468); I deferred that because we may need to do some general changes in the area (see #10802). So we can handle both at the same time later if we want (but should at least update the issue and put a TODO here).